File: backend/routes/shops.py
"""
Route per gestione negozi
"""
from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel
from typing import Optional
from uuid import UUID
from supabase import Client
from backend.database import get_supabase
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/shops", tags=["negozi"])


class ShopCreate(BaseModel):
    owner_id: UUID
    name: str
    description: Optional[str] = None
    address: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    website: Optional[str] = None


class ShopUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    address: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    website: Optional[str] = None


@router.get("/")
async def list_shops(
    owner_id: Optional[UUID] = None,
    supabase: Client = Depends(get_supabase)
):
    """Lista negozi con filtri opzionali"""
    try:
        query = supabase.table("shops").select("*")
        
        if owner_id:
            query = query.eq("owner_id", str(owner_id))
        
        result = query.execute()
        return {
            "shops": result.data,
            "count": len(result.data)
        }
    except Exception as e:
        logger.error(f"Errore lista negozi: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante il recupero negozi: {str(e)}"
        )


@router.get("/{shop_id}")
async def get_shop(shop_id: UUID, supabase: Client = Depends(get_supabase)):
    """Ottieni dettagli di un negozio"""
    try:
        result = supabase.table("shops").select("*").eq("id", str(shop_id)).execute()
        
        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Negozio non trovato"
            )
        
        return {"shop": result.data[0]}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore recupero negozio: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante il recupero negozio: {str(e)}"
        )


@router.post("/", status_code=status.HTTP_201_CREATED)
async def create_shop(
    shop: ShopCreate,
    current_user: dict = Depends(get_current_user),
    supabase: Client = Depends(get_supabase)
):
    """Crea un nuovo negozio"""
    try:
        # Solo negozianti possono creare negozi
        if current_user["role"] != "negoziante":
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Solo i negozianti possono creare negozi"
            )
        
        shop_data = shop.dict()
        # Usa l'ID dell'utente corrente come owner
        shop_data["owner_id"] = current_user["id"]
        
        result = supabase.table("shops").insert(shop_data).execute()
        
        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Errore durante la creazione del negozio"
            )
        
        return {
            "message": "Negozio creato con successo",
            "shop": result.data[0]
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore creazione negozio: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Errore durante la creazione del negozio: {str(e)}"
        )


@router.put("/{shop_id}")
async def update_shop(shop_id: UUID, shop: ShopUpdate, supabase: Client = Depends(get_supabase)):
    """Aggiorna un negozio"""
    try:
        updates = shop.dict(exclude_unset=True)
        
        if not updates:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Nessun campo da aggiornare"
            )
        
        result = supabase.table("shops").update(updates).eq("id", str(shop_id)).execute()
        
        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Negozio non trovato"
            )
        
        return {
            "message": "Negozio aggiornato con successo",
            "shop": result.data[0]
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Errore aggiornamento negozio: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante l'aggiornamento del negozio: {str(e)}"
        )


@router.delete("/{shop_id}")
async def delete_shop(shop_id: UUID, supabase: Client = Depends(get_supabase)):
    """Elimina un negozio"""
    try:
        result = supabase.table("shops").delete().eq("id", str(shop_id)).execute()
        
        return {
            "message": "Negozio eliminato con successo",
            "shop_id": str(shop_id)
        }
    except Exception as e:
        logger.error(f"Errore eliminazione negozio: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Errore durante l'eliminazione del negozio: {str(e)}"
        )

File: frontend/pages/shops.js
// Gestione negozi per negozianti
// Usa namespace per evitare conflitti se script caricato più volte
(function() {
    'use strict';
    
    // Verifica se già inizializzato
    if (window.shopsPageInitialized) {
        return;
    }
    window.shopsPageInitialized = true;
    
// Usa API_BASE_URL globale da app.js o fallback
const getApiBaseUrl = () => window.API_BASE_URL || window.CONFIG?.API_BASE_URL || 'http://localhost:8000';

// Helper per chiamate API
async function apiCall(endpoint, options = {}) {
    const token = localStorage.getItem('crm_token');
    const headers = {
        'Content-Type': 'application/json',
        ...options.headers
    };
    
    if (token) {
        headers['Authorization'] = `Bearer ${token}`;
    }
    
    try {
        const response = await fetch(`${getApiBaseUrl()}${endpoint}`, {
            headers,
            ...options
        });
        
        if (!response.ok) {
            const error = await response.json().catch(() => ({ detail: 'Errore sconosciuto' }));
            throw new Error(error.detail || `HTTP ${response.status}`);
        }
        
        return await response.json();
    } catch (error) {
        console.error('API Error:', error);
        throw error;
    }
}

// Carica lista negozi
async function loadShops() {
    try {
        const user = JSON.parse(localStorage.getItem('crm_user') || '{}');
        if (!user.id) {
            console.error('Utente non autenticato');
            return [];
        }
        
        const data = await apiCall(`/api/shops/?owner_id=${user.id}`);
        renderShops(data.shops || []);
        return data.shops || [];
    } catch (error) {
        console.error('Errore caricamento negozi:', error);
        document.getElementById('shops-list').innerHTML = 
            `<p class="error">Errore nel caricamento negozi: ${error.message}</p>`;
        return [];
    }
}

// Renderizza lista negozi
function renderShops(shops) {
    const container = document.getElementById('shops-list');
    if (!container) return;
    
    if (shops.length === 0) {
        container.innerHTML = '<p class="empty-state">Nessun negozio trovato. Crea il tuo primo negozio!</p>';
        return;
    }
    
    container.innerHTML = shops.map(shop => `
        <div class="shop-card">
            <h4>${shop.name || 'Negozio senza nome'}</h4>
            ${shop.description ? `<p>${shop.description}</p>` : ''}
            ${shop.address ? `<p>📍 ${shop.address}</p>` : ''}
            ${shop.phone ? `<p>📞 ${shop.phone}</p>` : ''}
            ${shop.email ? `<p>✉️ ${shop.email}</p>` : ''}
            <div class="shop-actions">
                <button class="btn btn-secondary" onclick="editShop('${shop.id}')">Modifica</button>
                <button class="btn btn-danger" onclick="deleteShop('${shop.id}')">Elimina</button>
            </div>
        </div>
    `).join('');
}

// Mostra form creazione negozio
function showCreateShopForm() {
    const user = JSON.parse(localStorage.getItem('crm_user') || '{}');
    if (!user.id) {
        window.showError('Devi essere autenticato per creare un negozio');
        return;
    }
    
    const modal = document.createElement('div');
    modal.className = 'modal';
    modal.id = 'create-shop-modal';
    modal.innerHTML = `
        <div class="modal-content">
            <div class="modal-header">
                <h3>Nuovo Negozio</h3>
                <button class="close-btn" onclick="closeModal('create-shop-modal')">&times;</button>
            </div>
            <div class="modal-body">
                <form id="create-shop-form">
                    <div class="form-group">
                        <label for="shop-name">Nome Negozio *</label>
                        <input type="text" id="shop-name" required>
                    </div>
                    <div class="form-group">
                        <label for="shop-description">Descrizione</label>
                        <textarea id="shop-description" rows="3"></textarea>
                    </div>
                    <div class="form-group">
                        <label for="shop-address">Indirizzo</label>
                        <input type="text" id="shop-address">
                    </div>
                    <div class="form-group">
                        <label for="shop-phone">Telefono</label>
                        <input type="tel" id="shop-phone">
                    </div>
                    <div class="form-group">
                        <label for="shop-email">Email</label>
                        <input type="email" id="shop-email">
                    </div>
                    <div class="form-group">
                        <label for="shop-website">Sito Web</label>
                        <input type="url" id="shop-website">
                    </div>
                    <div class="form-actions">
                        <button type="button" class="btn btn-secondary" onclick="closeModal('create-shop-modal')">Annulla</button>
                        <button type="submit" class="btn btn-primary">Crea Negozio</button>
                    </div>
                </form>
            </div>
        </div>
    `;
    
    document.body.appendChild(modal);
    
    document.getElementById('create-shop-form').addEventListener('submit', async (e) => {
        e.preventDefault();
        await createShop(user.id);
    });
}

// Crea nuovo negozio
async function createShop(ownerId) {
    try {
        const shopData = {
            owner_id: ownerId,
            name: document.getElementById('shop-name').value,
            description: document.getElementById('shop-description').value || null,
            address: document.getElementById('shop-address').value || null,
            phone: document.getElementById('shop-phone').value || null,
            email: document.getElementById('shop-email').value || null,
            website: document.getElementById('shop-website').value || null
        };
        
        await apiCall('/api/shops/', {
            method: 'POST',
            body: JSON.stringify(shopData)
        });
        
        window.showSuccess('Negozio creato con successo!');
        closeModal('create-shop-modal');
        loadShops();
    } catch (error) {
        window.showError('Errore nella creazione del negozio: ' + error.message);
    }
}

// Modifica negozio
function editShop(shopId) {
    // TODO: Implementare modifica negozio
    window.showError('Funzionalità di modifica non ancora implementata');
}

// Elimina negozio
async function deleteShop(shopId) {
    if (!confirm('Sei sicuro di voler eliminare questo negozio?')) {
        return;
    }
    
    try {
        await apiCall(`/api/shops/${shopId}`, {
            method: 'DELETE'
        });
        
        window.showSuccess('Negozio eliminato con successo!');
        loadShops();
    } catch (error) {
        window.showError('Errore nell\'eliminazione del negozio: ' + error.message);
    }
}

// Helper per chiudere modale
function closeModal(modalId) {
    const modal = document.getElementById(modalId);
    if (modal) {
        modal.remove();
    }
}

// Esporta funzioni per uso globale
window.loadShops = loadShops;
window.showCreateShopForm = showCreateShopForm;
window.editShop = editShop;
window.deleteShop = deleteShop;
window.closeModal = closeModal;

// Carica negozi all'avvio se siamo nella dashboard
if (window.location.pathname === '/dashboard') {
    setTimeout(() => {
        const shopsTab = document.getElementById('shop-tab-shops');
        if (shopsTab && shopsTab.classList.contains('active')) {
            loadShops();
        }
    }, 500);
}

